test/HsLuaSpec.hs

```diff
@@ -69,3 +83,55 @@ nulString =
     str' <- tostring l 1
     close l
     assertEqual "Popped string is different than what's pushed" str (B.unpack str')
+
+
+-----
+-- Random Quickcheck testing for StackValue instances
+-----
+
+-- Bools
+prop_bool :: Bool -> Property
+prop_bool = testAllStackValueInstance
+-- Ints
+prop_int :: Int -> Property
+prop_int = testAllStackValueInstance
+-- Bytestrings
+prop_bytestring :: ByteString -> Property
+prop_bytestring = testAllStackValueInstance
+-- Lists of bools
+prop_lists_bool :: [Bool] -> Property
+prop_lists_bool = testAllStackValueInstance
+-- Lists of ints
+prop_lists_int :: [Int] -> Property
+prop_lists_int = testAllStackValueInstance
+-- Lists of bytestrings
+prop_lists_bytestring :: [ByteString] -> Property
+prop_lists_bytestring = testAllStackValueInstance
+
+-- Check that the StackValue instance for a datatype works
+testStackValueInstance :: (Eq t, StackValue t) => t -> Property
+testStackValueInstance xs = QM.monadicIO $ do
+  x <- QM.run $ do
+         l <- newstate
+         push l xs
+         peek l (-1)
+  QM.assert $ xs == (fromJust x)
+
+-- Check that pushing/popping multiple times works
+testMultiStackValueInstance :: (Eq t, StackValue t) => t -> Property
+testMultiStackValueInstance xs = QM.monadicIO $ do
+  (x1,x2) <- QM.run $ do
+             l <- newstate
+             push l xs
+             push l ()
+             push l ()
+             push l xs
+             x1 <- peek l (-1)
+             x2 <- peek l (-4)
+             return (x1, x2)
+  QM.assert $ xs == (fromJust x1)
+  QM.assert $ xs == (fromJust x2)
```

Comment:
If we introduce a bug in the future, `fromJust` will fail with horrible error message because it's a partial function that we expect to work fine here, assuming that we don't have a bug. Instead, wrapping `xs` with `Just` would be better here, because in that case we'd get a nice error message if we introduce a bug.

i.e.

```haskell
QM.assert $ Just xs == fromJust x1
```

Same for the next line.